### ISSUE TITLE
can't search for some genera or families in 'choose a genus/family' bar

### ISSUE THREAD

**Issue by tmesaglio** [traitecoevo/infinitylists#62]
1. Set coordinates to -30.384972, 153.011733, choose a 50 km radius and search for all genera
2. Returns thousands of records, and if you search for the genus Sparganium within the species records tab search bar, you get one result:
![image](https://github.com/traitecoevo/infinitylists/assets/48739968/93c64995-a5b6-4cc0-a2e3-c8b7b6fa4a93)
3. Now click in the 'choose a genus' bar on the left, and start to type in Sparganium
4. There is no match, even though it's clearly a legitimate choice
5. This is almost certainly related to the issue at https://github.com/traitecoevo/infinitylists/issues/60

**Comment by wcornwell** [comment]
tough one. any idea if this missing genera thing is only for the select a lat/long? Or does it should up for the pre-loaded kml's also?

**Comment by tmesaglio** [comment]
pre-loaded KMLs too

**Comment by wcornwell** [comment]
what's an example?

**Comment by wcornwell** [comment]
i think i found a bug in the circle code that's not updating the genus/family lists as you move around and you get the last preloaded kml genus list, but if it's in the preloaded kml's too then there is a deeper issues.

**Comment by tmesaglio** [comment]
Sparganium is unsearchable for all of them

**Comment by wcornwell** [comment]
we have to do some kind of spatial subset of genera or else it crashes to the whole app...

**Comment by wcornwell** [comment]
not working perfectly yet, but can find Sparganium now: 
<img width="1448" alt="Screenshot 2023-09-19 at 5 08 32 pm" src="https://github.com/traitecoevo/infinitylists/assets/5534161/6265dac1-59f1-4282-b919-aed903a6c040">

**Comment by tmesaglio** [comment]
I don't know if this solves the problem at all, but what if the only option in the dropdown menu was 'all', and then you could also search in it to retrieve a single genus? Just remove all the other genera from being displayed, would that help? Same thing for family menu

**Comment by wcornwell** [comment]
not sure. let's discuss at the next meeting

**Comment by wcornwell** [comment]
chatGPT's thoughts on this problem: 
<img width="255" alt="Screenshot 2023-09-20 at 9 31 41 am" src="https://github.com/traitecoevo/infinitylists/assets/5534161/9dde864e-13dc-4510-9c6f-575b81965068">

**Comment by wcornwell** [comment]
right now we're doing (4), and we definitely have the problem that the robot mentions at the end....

**Comment by tmesaglio** [comment]
when I just ran the app locally, this is now broken again. For some of the preloaded places, I can't search for any genera or families

**Comment by wcornwell** [comment]
arggg, it's fine on my mac, but i can get the bug on the windows machine...

**Comment by tmesaglio** [comment]
it's weird as well because I tested for one preloaded place, couldn't search for anthing -> swapped to another place, **could** search for a genus that I tested --> swapped back to the first place and that single genus was now on the list along with 'all'. So something convoluted going on

**Comment by wcornwell** [comment]
yup, i have the same--something not updating properly when you switch places.

**Comment by wcornwell** [comment]
The issue you're facing with `updateSelectizeInput` not producing any options because `intersect_data()` isn't done yet is likely a synchronization problem between reactive dependencies.

When a user event triggers changes, some reactive computations may not finish before others are executed, leading to the described behavior. Here's a solution using `reactiveVal` to ensure synchronization:

1. **Introduce a Reactive Value**: Use `reactiveVal` to hold the state for the select inputs. This ensures that `updateSelectizeInput` only updates once the required computations are done.

```r
# Initialize a reactive value
data_ready <- reactiveVal(FALSE)
```

2. **Update `data_ready` after `intersect_data()` computation**: Within the reactive block for `intersect_data()`, set `data_ready` to `TRUE` once the computation finishes.

```r
intersect_data <- reactive({
    #... your existing code ...

    # Set data_ready to true once the computation finishes
    data_ready(TRUE)
    return(data)
})
```

3. **Observe `data_ready` and update inputs**: Instead of directly observing the `combined_input()`, observe the `data_ready` value.

```r
observe({
    if (data_ready()) {
        # Reset the reactive value
        data_ready(FALSE)
        
        updateSelectizeInput(
            session,
            "taxa_genus",
            selected = "All",
            choices = update_genus_choices(input$place),
            server = TRUE
        )
        
        updateSelectizeInput(
            session,
            "taxa_family",
            choices = update_family_choices(input$place),
            selected = "All",
            server = TRUE
        )
    }
})
```

This way, the input updates will only occur after `intersect_data()` has computed its result.

However, another way to handle this is by directly integrating the `updateSelectizeInput` within the `intersect_data` reactive, right after your data manipulations, ensuring that they're executed right after your data is ready. This might lead to a more synchronous behavior but could clutter your reactive definitions. The approach with `reactiveVal` is more modular and explicit about the synchronization step.

Lastly, ensure you don't have other reactivity chains that might cause premature triggers. The Shiny reactive system is quite powerful, but it's essential to be mindful of the order and dependencies between reactive elements.

**Comment by wcornwell** [comment]
closing for now...
